pygeos/lib/geom.pyx

```diff
@@ -0,0 +1,83 @@
+from cpython cimport PyObject
+cimport cython
+
+import numpy as np
+cimport numpy as np
+
+
+from pygeos.lib.geos_wrapper cimport *
+from pygeos.lib.pygeos_wrapper cimport *
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef geos_get_num_geometries(object[:] array):
+    cdef unsigned int i = 0
+    cdef GEOSContextHandle_t geos_handle = get_geos_handle()
+    cdef GEOSGeometry *geom = NULL
+
+    counts = np.zeros(shape=(array.size), dtype=np.intp)
+    cdef np.intp_t [:] counts_view = counts[:]
+
+    for i in range(array.size):
+        if PyGEOSGetGEOSGeom(<GeometryObject *>array[i], &geom) == 0:
+            raise TypeError("One of the arguments is of incorrect type. Please provide "
+            "only Geometry objects.")
+
+        if geom == NULL:
+            continue
+
+        counts_view[i] = GEOSGetNumGeometries_r(geos_handle, geom)
+
+    return counts
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def get_parts(object[:] array):
+    import_pygeos_core_api()
```

Comment:
This is the function that **must** be called in each top-level function in a Cython extension that uses the pygeos C API.

Reply:
I tried that initially.  It causes circular import issues with the way things are structured now.

As I understand it:
`import_pygeos_core_api()` triggers a module import from `pygeos.lib.core`, which traverses both `pygeos/__init__.py` and `pygeos/lib/__init__.py`, and the latter then imports from `pygeos/lib/geom.py` to promote its functions to the top level in `pygeos.lib` so that we can use them without needing to know that they come from a specific Cython module.

The same issue would be present if we wanted to pull anything from a Cython module that uses pygeos C API into the top level `__init__.py`, which we may well want to do in later PRs - which is why I picked the current approach.


There are a couple ways around this:
1) defer `import_pygeos_core_api` until a given top level function is first called (as here)
2) explicitly import from the specific Cython module in our Python wrappers (e.g., `geometry.py`)

Note: not all functions in a given Cython module may need to use the C API.

Maybe there is another way around the circular import issue?

Reply:
Right, I see the issue. Thanks for explaining. Could you place this explanation into the code as annotation?